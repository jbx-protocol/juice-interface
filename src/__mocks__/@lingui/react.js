const Trans = function ({ message, values }) {
    let replacedMessage = message
    for (const key in values) {
      // eslint-disable-next-line no-prototype-builtins
      if (values.hasOwnProperty(key)) {
        const regex = new RegExp(`\\{${key}\\}`, 'g')
        replacedMessage = replacedMessage.replace(regex, values[key])
      }
    }
    return replacedMessage
  },
  I18nProvider = function ({ children }) {
    return children
  }
export { Trans, I18nProvider }
