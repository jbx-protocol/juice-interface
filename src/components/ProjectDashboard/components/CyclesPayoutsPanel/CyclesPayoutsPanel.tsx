export const CyclesPanel = () => {
  return <>Cycles</>
}
