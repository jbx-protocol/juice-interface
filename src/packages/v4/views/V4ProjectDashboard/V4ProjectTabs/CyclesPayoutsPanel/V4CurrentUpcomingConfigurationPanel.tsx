import { ConfigurationPanel } from 'components/Project/ProjectTabs/CyclesPayoutsTab/ConfigurationPanel'
import { useV4CurrentUpcomingConfigurationPanel } from './hooks/useV4CurrentUpcomingConfigurationPanel'

type V4CurrentUpcomingConfigurationPanelProps = {
  type: 'current' | 'upcoming'
}

export const V4CurrentUpcomingConfigurationPanel: React.FC<
V4CurrentUpcomingConfigurationPanelProps
> = ({ type }) => {
  const props = useV4CurrentUpcomingConfigurationPanel(type)
  return <ConfigurationPanel {...props} />
}
