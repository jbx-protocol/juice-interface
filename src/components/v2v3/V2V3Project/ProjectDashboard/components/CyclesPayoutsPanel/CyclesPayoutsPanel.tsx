import { Tab } from '@headlessui/react'
import { t } from '@lingui/macro'
import { useMemo } from 'react'
import { CurrentUpcomingSubPanel } from './components/CurrentUpcomingSubPanel'
import { CyclesTab } from './components/CyclesTab'
import { HistorySubPanel } from './components/HistorySubPanel'

export type CyclesSubPanel = {
  id: 'current' | 'upcoming' | 'history'
  name: string
}

export const CyclesPayoutsPanel = () => {
  const tabs: CyclesSubPanel[] = useMemo(
    () => [
      { id: 'current', name: t`Current` },
      { id: 'upcoming', name: t`Upcoming` },
      { id: 'history', name: t`History` },
    ],
    [],
  )
  return (
    <Tab.Group
      as="div"
      className="mx-auto flex w-full max-w-[596px] flex-col gap-5"
    >
      <div className="flex flex-col justify-between gap-4 md:flex-row md:items-center">
        <h2 className="mb-0 font-heading text-2xl font-medium">Cycle</h2>
        <Tab.List className="flex gap-2">
          {tabs.map(tab => (
            <CyclesTab key={tab.id} name={tab.name} />
          ))}
        </Tab.List>
      </div>
      <Tab.Panels>
        {tabs.map(tab => (
          <Tab.Panel key={tab.id} className="outline-none">
            {tab.id === 'history' ? (
              <HistorySubPanel />
            ) : (
              <CurrentUpcomingSubPanel id={tab.id} />
            )}
          </Tab.Panel>
        ))}
      </Tab.Panels>
    </Tab.Group>
  )
}
