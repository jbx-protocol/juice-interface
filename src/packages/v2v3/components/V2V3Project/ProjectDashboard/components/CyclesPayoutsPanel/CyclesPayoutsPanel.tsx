import { Tab } from '@headlessui/react'
import { t } from '@lingui/macro'
import { CyclesTab } from 'components/Project/ProjectTabs/CyclesPayoutsTab/CyclesTab'
import { useProjectContext } from '../../hooks/useProjectContext'
import { CurrentUpcomingSubPanel } from './components/CurrentUpcomingSubPanel'
import { HistorySubPanel } from './components/HistorySubPanel'

type CyclesSubPanel = {
  id: 'current' | 'upcoming' | 'history'
  name: string
}

export const CyclesPayoutsPanel = () => {
  const { fundingCycle } = useProjectContext()

  const tabs: CyclesSubPanel[] = fundingCycle ? [
    // Don't show the current tab if there is no current cycle
    fundingCycle.number.gt(0) && { id: 'current', name: t`Current` },
    { id: 'upcoming', name: t`Upcoming` },
    { id: 'history', name: t`History` },
  ].filter(Boolean) as CyclesSubPanel[] : []

  return (
    <Tab.Group as="div" className="mx-auto flex w-full flex-col gap-5">
      <div className="flex flex-col justify-between gap-4 md:flex-row md:items-center">
        <h2 className="mb-0 font-heading text-2xl font-medium">Cycle</h2>
        <Tab.List className="flex gap-2">
          {tabs.map(tab => (
            <CyclesTab key={tab.id} name={tab.name} />
          ))}
        </Tab.List>
      </div>
      <Tab.Panels>
        {tabs.map(tab => (
          <Tab.Panel key={tab.id} className="outline-none">
            {tab.id === 'history' ? (
              <HistorySubPanel />
            ) : (
              <CurrentUpcomingSubPanel id={tab.id} />
            )}
          </Tab.Panel>
        ))}
      </Tab.Panels>
    </Tab.Group>
  )
}
