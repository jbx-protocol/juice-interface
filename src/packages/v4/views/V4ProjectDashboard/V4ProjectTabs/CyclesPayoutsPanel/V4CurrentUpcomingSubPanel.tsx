import { InformationCircleIcon } from '@heroicons/react/24/outline'
import { Trans, t } from '@lingui/macro'
import { currentCycleRemainingLengthTooltip } from 'components/Project/ProjectTabs/CyclesPayoutsTab/CyclesPanelTooltips'
import { UpcomingCycleChangesCallout } from 'components/Project/ProjectTabs/CyclesPayoutsTab/UpcomingCycleChangesCallout'
import { TitleDescriptionDisplayCard } from 'components/Project/ProjectTabs/TitleDescriptionDisplayCard'
import { useMemo } from 'react'
import { twMerge } from 'tailwind-merge'
import { useV4CurrentUpcomingSubPanel } from '../../hooks/useV4CurrentUpcomingSubPanel'
import { V4ConfigurationDisplayCard } from './V4ConfigurationDisplayCard'
import { useV4UpcomingRulesetHasChanges } from './hooks/useV4UpcomingRulesetHasChanges'

const CYCLE_NUMBER_INDEX = 0
const STATUS_INDEX = 1
const CYCLE_LENGTH_INDEX = 2

export const V4CurrentUpcomingSubPanel = ({
  id,
}: {
  id: 'current' | 'upcoming'
}) => {
  const info = useV4CurrentUpcomingSubPanel(id)
  const { hasChanges, loading } = useV4UpcomingRulesetHasChanges()

  const topPanelsInfo = useMemo(() => {
    const topPanelInfo = [
      {
        title: t`Ruleset #`,
        value: info.rulesetNumber,
      },
      {
        title: t`Status`,
        value: info.status,
      },
    ]
    if (info.type === 'current') {
      topPanelInfo.push({
        title: t`Remaining time`,
        value: info.remainingTime,
      })
    }
    if (info.type === 'upcoming') {
      topPanelInfo.push({
        title: t`Ruleset duration`,
        value: info.rulesetLength,
      })
    }
    return topPanelInfo
  }, [
    info.rulesetLength,
    info.rulesetNumber,
    info.remainingTime,
    info.status,
    info.type,
  ])

  const rulesetLengthTooltip =
    info.type === 'current' ? currentCycleRemainingLengthTooltip : undefined

  const rulesetLengthValue = topPanelsInfo[CYCLE_LENGTH_INDEX].value?.toString()

  const rulesetStatusTooltip = info.currentRulesetUnlocked ? (
    <Trans>The project's rules are unlocked and can change at any time.</Trans>
  ) : (
    <Trans>
      This project's rules will be locked in place for {rulesetLengthValue}.
    </Trans>
  )

  const hasNoUpcomingRuleset =
    info.type === 'upcoming' &&
    info.currentRulesetUnlocked &&
    /**
     * Always show 'upcoming' tab if it's FC 1
     * (which happens when Scheduled Launch is used,
     * mustStartAtOrAfter is in the future)
     */
    info.rulesetNumber !== 1 &&
    !info.hasPendingConfiguration

  if (hasNoUpcomingRuleset) {
    return (
      <div>
        <div
          className={twMerge(
            'flex items-center gap-2 rounded-lg py-2 px-3.5 text-sm font-medium shadow-sm',
            'bg-smoke-75 text-smoke-700 dark:bg-slate-400 dark:text-slate-200',
          )}
        >
          <InformationCircleIcon className="h-5 w-5" />
          <Trans>
            This project has no upcoming ruleset. Its rules can change at any
            time.
          </Trans>
        </div>
      </div>
    )
  }

  const upcomingRulesetChangesCalloutText = hasChanges
    ? t`This ruleset has upcoming changes`
    : t`This ruleset has no upcoming changes`

  return (
    <div>
      <div className="flex flex-col gap-4">
        {id === 'upcoming' && (
          <UpcomingCycleChangesCallout
            text={upcomingRulesetChangesCalloutText}
            loading={loading}
            hasChanges={hasChanges}
          />
        )}

        <div className="grid grid-cols-2 gap-4 md:flex">
          <TitleDescriptionDisplayCard
            className="w-full md:max-w-[127px]"
            title={topPanelsInfo[CYCLE_NUMBER_INDEX].title}
            description={
              topPanelsInfo[CYCLE_NUMBER_INDEX].value?.toString() ?? <Skeleton />
            }
          />
          <TitleDescriptionDisplayCard
            className="w-full md:w-fit"
            title={topPanelsInfo[STATUS_INDEX].title}
            description={
              topPanelsInfo[STATUS_INDEX].value?.toString() ?? <Skeleton className="w-22" />
            }
            tooltip={rulesetStatusTooltip}
          />
          <TitleDescriptionDisplayCard
            className="col-span-2 md:flex-1"
            title={topPanelsInfo[CYCLE_LENGTH_INDEX].title}
            description={
              topPanelsInfo[CYCLE_LENGTH_INDEX].value?.toString() ?? (
                <Skeleton className="w-40" />
              )
            }
            tooltip={rulesetLengthTooltip}
          />
        </div>
        <V4ConfigurationDisplayCard type={info.type} />
      </div>

      {/* <PayoutsSubPanel className="mt-12" type={id} /> */}
    </div>
  )
}

const Skeleton = ({ className }: { className?: string }) => (
  <div
    className={twMerge(
      'h-8 w-12 animate-pulse rounded-lg bg-smoke-200 dark:bg-slate-500',
      className,
    )}
  />
)
