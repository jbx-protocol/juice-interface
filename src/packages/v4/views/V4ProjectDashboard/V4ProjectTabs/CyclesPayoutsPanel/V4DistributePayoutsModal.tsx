import { t, Trans } from '@lingui/macro'
import { Form } from 'antd'
import InputAccessoryButton from 'components/buttons/InputAccessoryButton'
import { Callout } from 'components/Callout/Callout'
import FormattedNumberInput from 'components/inputs/FormattedNumberInput'
import TransactionModal from 'components/modals/TransactionModal'
import { FEES_EXPLANATION } from 'components/strings'
import { usePayoutLimit } from 'packages/v4/hooks/usePayoutLimit'
import { useUsedPayoutLimitOf } from 'packages/v4/hooks/useUsedPayoutLimitOf'
import { useV4BalanceOfNativeTerminal } from 'packages/v4/hooks/useV4BalanceOfNativeTerminal'
import { V4CurrencyName } from 'packages/v4/utils/currency'
import { useEffect, useState } from 'react'
import { fromWad } from 'utils/format/formatNumber'
import { useV4CurrentPayoutSplits } from '../../../../hooks/useV4PayoutSplits'

export default function V4DistributePayoutsModal({
  open,
  onCancel,
  onConfirmed,
}: {
  open?: boolean
  onCancel?: VoidFunction
  onConfirmed?: VoidFunction
}) {
  const { splits: payoutSplits } = useV4CurrentPayoutSplits()
  const { data: usedPayoutLimit } = useUsedPayoutLimitOf()
  const { data: payoutLimit } = usePayoutLimit()
  const { data: balanceOfNativeTerminal } = useV4BalanceOfNativeTerminal()

  const payoutLimitAmount = payoutLimit?.amount
  const payoutLimitAmountCurrency = payoutLimit?.currency

  const [transactionPending, setTransactionPending] = useState<boolean>()
  const [loading, setLoading] = useState<boolean>()
  const [distributionAmount, setDistributionAmount] = useState<string>()

  // TODO: const v4DistributePayoutsTx = useV4DistributePayoutsTx()

  useEffect(() => {
    if (!payoutLimitAmount) return

    const unusedFunds = payoutLimitAmount - (usedPayoutLimit ?? 0n) ?? 0n
    const distributable = balanceOfNativeTerminal && balanceOfNativeTerminal > unusedFunds
      ? unusedFunds
      : 0n

    setDistributionAmount(fromWad(distributable))
  }, [
    balanceOfNativeTerminal,
    payoutLimitAmount,
    usedPayoutLimit,
  ])

  async function executeDistributePayoutsTx() {
    if (!payoutLimitAmountCurrency || !distributionAmount) return

    setLoading(true)

    const txSuccessful = true
    // TODO: const txSuccessful = await v4DistributePayoutsTx(
    //   {
    //     amount: distributionAmount,
    //     currency: payoutLimitAmountCurrency,
    //   },
    //   {
    //     onDone: () => {
    //       setTransactionPending(true)
    //     },
    //     onConfirmed: () => {
    //       setLoading(false)
    //       setTransactionPending(false)
    //       onConfirmed?.()
    //     },
    //   },
    // )

    if (!txSuccessful) {
      setLoading(false)
      setTransactionPending(false)
    }
  }

  const unusedFunds = payoutLimitAmount ?
    payoutLimitAmount - (usedPayoutLimit ?? 0n) : 0n

  const distributable = balanceOfNativeTerminal && balanceOfNativeTerminal > unusedFunds
    ? unusedFunds
    : 0n

  const currencyName = V4CurrencyName(payoutLimitAmountCurrency)

  return (
    <TransactionModal
      title={<Trans>Send payouts</Trans>}
      open={open}
      onOk={executeDistributePayoutsTx}
      onCancel={() => {
        setDistributionAmount(undefined)
        onCancel?.()
      }}
      okButtonProps={{
        disabled: !distributionAmount || distributionAmount === '0',
      }}
      confirmLoading={loading}
      transactionPending={transactionPending}
      okText={t`Send payouts`}
      connectWalletText={t`Connect wallet to send payouts`}
      width={640}
      className="top-[40px]"
    >
      <div className="flex flex-col gap-6">
        <Form layout="vertical">
          <Form.Item
            className="mb-0"
            label={<Trans>Amount to pay out</Trans>}
            extra={<Trans>Recipients will recieve payouts in ETH</Trans>}
          >
            <FormattedNumberInput
              placeholder="0"
              value={distributionAmount}
              onChange={value => setDistributionAmount(value)}
              min={0}
              accessory={
                <div className="flex items-center">
                  <span className="mr-2 text-black dark:text-slate-100">
                    {currencyName}
                  </span>
                  <InputAccessoryButton
                    content={<Trans>MAX</Trans>}
                    onClick={() =>
                      setDistributionAmount(fromWad(distributable))
                    }
                  />
                </div>
              }
            />
          </Form.Item>
        </Form>
        <div>
          <div className="text-md mb-2 font-medium">
            <Trans>Payout recipients</Trans>
          </div>

          {payoutSplits?.length === 0 ? (
            <Callout.Info className="mb-4">
              <Trans>
                The project owner is the only payout recipient. Any ETH paid out
                this cycle will go to them.
              </Trans>
            </Callout.Info>
          ) : null}

          <div className="max-h-[33vh] overflow-y-auto">
            {/* <PayoutsTable
              payoutSplits={payoutSplits ?? []}
              currency={currencyName}
              payoutLimitAmount={parseFloat(distributionAmount ?? '0')} // payoutLimitAmount is the amount to distribute in the instance.
              hideHeader
              showAvatars
            /> */}
          </div>
        </div>

        <Callout.Info>{FEES_EXPLANATION}</Callout.Info>
      </div>
    </TransactionModal>
  )
}
