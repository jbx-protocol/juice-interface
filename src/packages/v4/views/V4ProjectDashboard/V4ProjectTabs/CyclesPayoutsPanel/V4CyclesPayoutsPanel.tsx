import { Tab } from '@headlessui/react'
import { t } from '@lingui/macro'
import { CyclesTab } from 'components/Project/ProjectTabs/CyclesPayoutsTab/CyclesTab'
import { RulesetCountdownProvider } from 'packages/v4/contexts/RulesetCountdownProvider'
import { useMemo } from 'react'
import { V4CurrentUpcomingSubPanel } from './V4CurrentUpcomingSubPanel'

type V4CyclesSubPanel = {
  id: 'current' | 'upcoming' | 'history'
  name: string
}

export const V4CyclesPayoutsPanel = () => {
  const tabs: V4CyclesSubPanel[] = useMemo(
    () => [
      { id: 'current', name: t`Current` },
      { id: 'upcoming', name: t`Upcoming` },
      // { id: 'history', name: t`History` },
    ],
    [],
  )
  return (
    <Tab.Group as="div" className="mx-auto flex w-full flex-col gap-5">
      <div className="flex flex-col justify-between gap-4 md:flex-row md:items-center">
        <h2 className="mb-0 font-heading text-2xl font-medium">Cycle</h2>
        <Tab.List className="flex gap-2">
          {tabs.map(tab => (
            <CyclesTab key={tab.id} name={tab.name} />
          ))}
        </Tab.List>
      </div>
      <Tab.Panels>
        {tabs.map(tab => (
          <Tab.Panel key={tab.id} className="outline-none">
            {tab.id === 'history' ? (
              <></> //<HistorySubPanel />
            ) : (
              <RulesetCountdownProvider>
                <V4CurrentUpcomingSubPanel id={tab.id} />
              </RulesetCountdownProvider>
            )}
          </Tab.Panel>
        ))}
      </Tab.Panels>
    </Tab.Group>
  )
}
