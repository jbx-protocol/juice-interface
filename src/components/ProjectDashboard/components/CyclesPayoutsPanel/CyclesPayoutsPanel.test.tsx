/**
 * @jest-environment jsdom
 */
import { render } from '@testing-library/react'
import { CyclesPayoutsPanel } from './CyclesPayoutsPanel'

describe('CyclesPayoutsPanel', () => {
  it('renders', () => {
    const { container } = render(<CyclesPayoutsPanel />)
    expect(container).toMatchSnapshot()
  })
})
