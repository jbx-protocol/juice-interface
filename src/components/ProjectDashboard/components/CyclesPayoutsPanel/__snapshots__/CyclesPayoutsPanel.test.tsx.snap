// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CyclesPayoutsPanel renders 1`] = `
<div>
  <div
    class="mx-auto flex w-full max-w-[596px] flex-col gap-5"
  >
    <div
      class="flex flex-col justify-between gap-4 md:flex-row md:items-center"
    >
      <h2
        class="mb-0 font-heading text-2xl font-medium"
      >
        Cycle
      </h2>
      <div
        aria-orientation="horizontal"
        class="flex gap-2"
        role="tablist"
      >
        <button
          aria-controls="headlessui-tabs-panel-:r3:"
          aria-selected="true"
          class="outline-none"
          data-headlessui-state="selected"
          id="headlessui-tabs-tab-:r0:"
          role="tab"
          tabindex="0"
          type="button"
        >
          <div
            class="rounded-2xl py-0.5 px-3 font-medium bg-smoke-700 text-grey-25 dark:bg-slate-400 dark:text-slate-50"
          >
            Current
          </div>
        </button>
        <button
          aria-controls="headlessui-tabs-panel-:r4:"
          aria-selected="false"
          class="outline-none"
          data-headlessui-state=""
          id="headlessui-tabs-tab-:r1:"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <div
            class="rounded-2xl bg-smoke-50 py-0.5 px-3 font-medium text-smoke-600 dark:bg-slate-700 dark:text-slate-200"
          >
            Upcoming
          </div>
        </button>
        <button
          aria-controls="headlessui-tabs-panel-:r5:"
          aria-selected="false"
          class="outline-none"
          data-headlessui-state=""
          id="headlessui-tabs-tab-:r2:"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <div
            class="rounded-2xl bg-smoke-50 py-0.5 px-3 font-medium text-smoke-600 dark:bg-slate-700 dark:text-slate-200"
          >
            History
          </div>
        </button>
      </div>
    </div>
    <div>
      <div
        aria-labelledby="headlessui-tabs-tab-:r0:"
        class="outline-none"
        data-headlessui-state="selected"
        id="headlessui-tabs-panel-:r3:"
        role="tabpanel"
        tabindex="0"
      >
        <div>
          CurrentUpcomingSubPanel
        </div>
      </div>
      <span
        aria-labelledby="headlessui-tabs-tab-:r1:"
        id="headlessui-tabs-panel-:r4:"
        role="tabpanel"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        tabindex="-1"
      />
      <span
        aria-labelledby="headlessui-tabs-tab-:r2:"
        id="headlessui-tabs-panel-:r5:"
        role="tabpanel"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;
